### PR TITLE
[BEAM-2405] Override to sink interface in batch dataflow

### PR DESCRIPTION
**Pull request by sb2nov**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
R: @chamikaramj  PTAL

We can deprecate the sink interface to not be user facing now as this is a unified model.

### CODE REVIEW

File: sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes.py [hunk 1]
```diff
@@ -83,12 +83,11 @@ def run(argv=None):
 
     # Write the output using a "Write" transform that has side effects.
     # pylint: disable=expression-not-assigned
-    counts | 'write' >> beam.io.Write(
-        beam.io.BigQuerySink(
-            known_args.output,
-            schema='month:INTEGER, tornado_count:INTEGER',
-            create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
-            write_disposition=beam.io.BigQueryDisposition.WRITE_TRUNCATE))
+    counts | 'write' >> beam.io.WriteToBigQuery(
```
Comment (robertwb) [inline sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes.py]: Capitalize Write?

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 2]
```diff
@@ -378,6 +379,23 @@ def run_Flatten(self, transform_node):
           PropertyNames.ENCODING: step.encoding,
           PropertyNames.OUTPUT_NAME: PropertyNames.OUT}])
 
+  def apply_WriteToBigQuery(self, transform, pcoll):
+    standard_options = pcoll.pipeline._options.view_as(StandardOptions)
+    if standard_options.streaming:
+      write_truncate = beam.io.BigQueryDisposition.WRITE_TRUNCATE
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Nit: Is this temporary just for line wrapping? If so, I'd just use ()'s in the if.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 3]
```diff
@@ -378,6 +379,23 @@ def run_Flatten(self, transform_node):
...
+      if transform.write_disposition == write_truncate:
+        raise RuntimeError('Can not use write truncation mode in streaming')
+      return self.apply_PTransform(transform, pcoll)
+    else:
+      return pcoll  | 'Write to BigQuery' >> beam.io.Write(
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: CamelCase labels.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 4]
```diff
@@ -378,6 +379,23 @@ def run_Flatten(self, transform_node):
           PropertyNames.ENCODING: step.encoding,
           PropertyNames.OUTPUT_NAME: PropertyNames.OUT}])
 
+  def apply_WriteToBigQuery(self, transform, pcoll):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Do we also want to override batch mode for the dataflow runner?

Comment (sb2nov) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: this is only overriding the batch mode for streaming we can keep using the streaming api so the ptransfrom should be okay.

Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: As I read the code, this is using the native bigquery sink for batch on direct runner. Should we be doing the same for the dataflow runner (i.e. using the native sink for batch)?

Comment (sb2nov) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: I'm a bit confused. This is the dataflow_runner.py file so we're only overriding the one in dataflow runner. We have always used the streaming api for the pipelines in the direct runner.

File: sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes.py (hunk 1 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes.py]: Can you also add a ReadFromBigQuery transform for consistency ?

Comment (sb2nov) [inline sdks/python/apache_beam/examples/cookbook/bigquery_tornadoes.py]: Sure, I'll do that in a followup PR

**Comment by robertwb** [comment]
On Tue, Jun 6, 2017 at 4:07 PM, Sourabh Bajaj <notifications@github.com>
wrote:

> *@sb2nov* commented on this pull request.
> ------------------------------
>
> In sdks/python/apache_beam/runners/dataflow/dataflow_runner.py
> <https://github.com/apache/beam/pull/3306#discussion_r120503749>:
>
> > @@ -378,6 +379,23 @@ def run_Flatten(self, transform_node):
>            PropertyNames.ENCODING: step.encoding,
>            PropertyNames.OUTPUT_NAME: PropertyNames.OUT}])
>
> +  def apply_WriteToBigQuery(self, transform, pcoll):
>
> I'm a bit confused. This is the dataflow_runner.py file so we're only
> overriding the one in dataflow runner. We have always used the streaming
> api for the pipelines in the direct runner.
>
Oh, yes, I was thinking this was direct runner for some reason. This is
good.

**Comment by sb2nov** [comment]
PTAL again

**Comment by chamikaramj** [comment]
LGTM. Will merge after tests pass.

**Comment by sb2nov** [comment]
Should be ready to merge now.

**Pull request closed without merging**